### DISCUSSION TITLE
toolkit: multitenancy question

### DISCUSSION THREAD

**Discussion by Perdjesk** [fluxcd/flux2#93] [Q&A]
The current architecture of this new approach seems to assume that the "toolkit" controllers are running globally for all tenants (i.e assuming tenant are owning namespaces). 

### Diagram might be misleading
https://github.com/fluxcd/toolkit/blob/master/docs/diagrams/tk-feature.png

In this diagram "CRDs" are mentioned in a "tenant" box which I assume are namespaces. Whereas the custom objects boxes "Sources" and "Kustomizations" are outside of the tenant boxes. Reading the docs it is my understanding that it should be the reverse. The "CustomResourceDefinitions" are cluster level objects and should be out of a tenant box given that a tenant might not have cluster level permission to create them in a multi-tenancy model. The custom objects "Sources" and "Kustomizations" should be represented in the tenant box given the CRDs have scope "namespaced" and thus those are the objects that will be created by tenants within a namespace. Is this reasoning make sense?
See:
https://github.com/fluxcd/source-controller/blob/master/config/crd/bases/source.fluxcd.io_gitrepositories.yaml#L30
https://github.com/fluxcd/kustomize-controller/blob/master/config/crd/bases/kustomize.fluxcd.io_kustomizations.yaml#L27 

### Tenant isolation with shared controllers
In the context of a tenant owning a namespace without cluster level access that want to deploy to a specific namespace with private repository. 
My assumption would be that the tenant would be responsible to create one "Source" namespaced object with credentials and one "Kustomization" namespaced object referencing this "Source" object and the controller enforcing the permissions whether the "Kustomization" namespaces object has access to referenced "Source" object. Assuming as well that a namespaced "Kustomization" object has by default no permission and must explicitly reference a "serviceAccount" that is namespaced in the same namespace.

However by reading the following docs it seems that it is the responsibility of the admin to handle this for the tenants given the the "Kustomization" controller default to cluster admin access and admin are responsible to create "Kustomization" object scoped to a namespace. That seems to put a lot of legwork on the admins to handle this isolation for all tenants.
https://toolkit.fluxcd.io/components/kustomize/kustomization/#role-based-access-control

Wouldn't the previously described pattern possible in order to provide isolation between tenants fully managed by tenants themselves instead of relying on admin?
Is there currently any restrictions on which "sourceRef" can be used by "Kustomization" objects?

**Comment by stefanprodan** [comment]
The way we setup the control plane with `tk install/bootstrap` is for clusters where admins have to register sources and setup kustomizations for each team. A cluster user shouldn't have access to the `gitops-system` for this to work. 

> That seems to put a lot of legwork on the admins to handle this isolation for all tenants.

Yes this is the intention, to give sole ownership of sources and reconcilers SA to cluster admins.

> Is there currently any restrictions on which "sourceRef" can be used by "Kustomization" objects?

Yes, a Kustomization can reference a source only it the same namespace and since we setup source and kustomize controllers to only watch the namespace where they are deployed e.g. `gitops-system`, a user can't take over the cluster even if they create source or kustomization objects in their own repos/namespaces since no controller will act on them.

> Wouldn't the previously described pattern possible in order to provide isolation between tenants fully managed by tenants themselves instead of relying on admin?

Yes this could be possible if we setup the control plane like so:
- make source/kustomize controllers watch all namespaces by removing the RUNTIME_NAMESPACE env var
- write an admission controller that injects/overrides the SA in a Kustomization for non-admin namespaces (relies on having an allow-list of admin namespaces and service accounts)

**Reply by Perdjesk** [reply]
The diagram from https://github.com/fluxcd/multi-tenancy is more self-explanatory what is handled by admin/tenants.

> 
> 
> The way we setup the control plane with `tk install/bootstrap` is for clusters where admins have to register sources and setup kustomizations for each team. A cluster user shouldn't have access to the `gitops-system` for this to work.
> 
> > That seems to put a lot of legwork on the admins to handle this isolation for all tenants.
> 
> Yes this is the intention, to give sole ownership of sources and reconcilers SA to cluster admins.
> 

That seems strange behavior at first hand to me. As K8s cluster admin I would not want to manage all the Sources for all the tenants. That would requires a lot of work and sync with all the tenants. 

> > Is there currently any restrictions on which "sourceRef" can be used by "Kustomization" objects?
> 
> Yes, a Kustomization can reference a source only it the same namespace and since we setup source and kustomize controllers to only watch the namespace where they are deployed e.g. `gitops-system`, a user can't take over the cluster even if they create source or kustomization objects in their own repos/namespaces since no controller will act on them.

What you describe is that the admin is as well responsible to manage all Kustomization resources and correctly matches them to namespaces SAs and sources. Meaning that admin becomes responsible to ensure that a tenant has only access to his owned Sources. Especially in the case of private repositories with credentials, which would requires tenant to explicitly give repositories credentials to cluster admins. From the point of view of tenants admin management of those resources could easily become a bottleneck and a painful process.
I do not grasp the motivation of requiring such workflow from the perspective of admin or tenants.

> 
> > Wouldn't the previously described pattern possible in order to provide isolation between tenants fully managed by tenants themselves instead of relying on admin?
> 
> Yes this could be possible if we setup the control plane like so:
> 
>     * make source/kustomize controllers watch all namespaces by removing the RUNTIME_NAMESPACE env var
> 
>     * write an admission controller that injects/overrides the SA in a Kustomization for non-admin namespaces (relies on having an allow-list of admin namespaces and service accounts)

In such scenario I would expect the tenant creating Kustomization resources to always explicitly specify a SA to use for the deployment which has the required permission for this deployment. However if the Kustomize controller can then watch all namespaces resources how to restrict a Kustomization namespaces resource to not access a Source from another namespace? Using RBAC on the SA scope to that namespace would be enough or Kustomization controller is using cluster wide SA to access Sources?

IMO and in my experience of K8s cluster multi-tenancy using namespaces, the management of Source/Kustomization by tenant would be the primary use case and the most wanted workflow. Tenant that are doing ContinuousDelivery on Kubernetes are owners of their repositories/sources as well and generally don't want to rely on someone else to be able to handle that for them. As well admin are only interested to manage things for tenant only if it brings value either in security/audit or compliance. 

What I could see the benefits of having cluster wide and admin managed Source is for shared Sources that are use by multiple tenants and that admin take care of curating for the tenants and is available to all. But that would be complementary of the first self-managed Source/Kustomization approach.

WDYT?

**Reply by stefanprodan** [reply]
> However if the Kustomize controller can then watch all namespaces resources how to restrict a Kustomization namespaces resource to not access a Source from another namespace?

A Kustomization can't reference a source from another namespace:

```go
	// Reference of the source where the kustomization file is.
	// +required
	SourceRef corev1.TypedLocalObjectReference `json:"sourceRef"`
```

> Admin are only interested to manage things for tenant only if it brings value either in security/audit or compliance.

That's the reason the default installation setups the control plane like I've described. A tenant would create a PR on the fleet repository to add their sources and kustomization, once approved and merged by an admin, the tenant has the freedom to manage their Kubernetes resources in their own repo.

>  In the case of private repositories with credentials, which would requires tenant to explicitly give repositories credentials to cluster admins.

The tenant could use sealed-secrets and the PR will not reveal the credentials, anyone can encrypt secrets but only the sealed-secrets controller running inside the cluster can decrypt them.

**Reply by Perdjesk** [reply]
> 
> A Kustomization can't reference a source from another namespace:
> 
> ```go
> 	// Reference of the source where the kustomization file is.
> 	// +required
> 	SourceRef corev1.TypedLocalObjectReference `json:"sourceRef"`
> ```
> 

That would be great if it was a ObjectReference instead allowing to reference Source that could be shared among tenants from a common namespace managed by admin and for which tenants have read access to. The flexibility might allow others use cases potentially.

> > Admin are only interested to manage things for tenant only if it brings value either in security/audit or compliance.
> 
> That's the reason the default installation setups the control plane like I've described. A tenant would create a PR on the fleet repository to add their sources and kustomization, once approved and merged by an admin, the tenant has the freedom to manage their Kubernetes resources in their own repo.

As a tenant of a K8s cluster that would not be super convenient to have to interact with a admin managed repository for that. The service main interface is the Kubernetes API and if possible it would be better to have tenant use it for setting up their CD pipepline via Source/Kustomizations resources instead of a supplementary Git/PR interface. I still don't fully get which benefits the admin get out of managing this even with a PR workflow. Moreover a shared repository even with encrypted secrets would disclose non-necessary information about tenant to other tenants. Not sure it is really wished to have tenant be able to see what other tenants are doing/using as Sources or Kustomization deployment details.  

I can see the benefits of both approach for different use cases. A small cluster with a few tenants that might know each other and don't mind sharing some stuff between them and information about them. With that an admin doing the heavy lifting of setting up the CD within one namespace for all those teams with maybe even a shared pattern of Sources/Kustomizations for different projects and teams then only interacting with their manifest repositories and requesting new Sources to their admin.
On the other hand a substantially bigger cluster with several dozens of tenants that are handling a wide variety of applications and that are interested to remain unaware of others tenants and vice-versa. With an admin team which has no knowledge of tenant workload or CD setup that want to provide a self-service for tenant to bootstrap CD pipeline based on the toolkit but not managed every tenant specificity and even less their private setup and credentials. 

Maybe it would be possible to accommodate both approach and let users decide which way they want. With experience of the latter case admin and tenants have more benefits with a clear separation of concerns and self-service capabilities using the K8s API.

**Reply by stefanprodan** [reply]
> That would be great if it was a ObjectReference instead allowing to reference Source that could be shared among tenants from a common namespace managed by admin and for which tenants have read access to. The flexibility might allow others use cases potentially.

That is something that we consider doing, it would improve collaboration when tenants don't share the fleet repo with the admins. https://github.com/fluxcd/kustomize-controller/issues/65

 > Maybe it would be possible to accommodate both approach and let users decide which way they want. With experience of the latter case admin and tenants have more benefits with a clear separation of concerns and self-service capabilities using the K8s API.

I think the toolkit should accommodate both approaches, with the latter my concern is that it requires a non-gitops workflow. If there is no central repository where admins review/approve sources, than this logic moves to admission controllers and sources/kustomizations would have to be created outside of git e.g. with kubectl apply by a team member.

**Reply by Perdjesk** [reply]
> I think the toolkit should accommodate both approaches, with the latter my concern is that it requires a non-gitops workflow. If there is no central repository where admins review/approve sources, than this logic moves to admission controllers and sources/kustomizations would have to be created outside of git e.g. with kubectl apply by a team member.

I see now the idea behind having this central repository in order to allow tenants to configure their CD pipeline with a GitOps approach. From the docs as well I get as well that the chicken or the egg dilemma of whom comes first between the toolkit components/control plane and the Git repository containing their manifest definitions is handled by the `tk` bootstrap CLI command. 
For Sources/Kustomizations per namespace handled by teams without an admin controlled central repository, the `tk` CLI could potentially be used as well by tenant to bootstrap their CD pipeline within a namespace in sync with a repository which would converge with an admin workflow used for the `tk` control plane. `tk` CLI would then be the only promoted hybrid GitOps solution for bootstrapping GitOps CD pipeline for admin and tenants.

Regarding the use case of admin reviewing/approving Sources objects cluster wide for all tenants. A central repository for Sources, controlled by an admin, is a quick-win solution which is restrictive given:
1. Admin can't set generic allow/deny policies but have to approves/review every PR to the repository.
2. Tenant must disclose private Sources credentials to admin explicitly (i.e even if encrypted using a admin's secret within the repository as you proposed). 
3. Tenants information disclosure. Tenants will gain knowledge about others tenants existence and CD/system configurations through the central repository. Maybe this one could be mitigated with a Git submodule per tenant within a central mono-repository.

To improve on those restrictions we could imagine decoupling the Sources objects and the reviewing/approving by introducing a SourcePolicy object that defines policies on Sources URLs for tenants (or any other Source attributes). The `tk` control plane would provide a controller with an admission webhook that is fed by SourcePolicy objects. SourcePolicy objects would be controlled by admin through a central repository open for PR by tenants. Tenants would be responsible to create Source objects in their own namespaces using the `tk` CLI to sync it with a Git repository. That would improve 1 and 2 limitations but 3 is still present.

Now taking the limitation 3 and thinking more globally in a context where admin wants to streamline the management of tenant configuration using a GitOps approach and provide a GitOps pattern for tenant to request admin managed K8s capabilities. For example tenant requesting resources quotas, network policies, RBAC/ABAC permissions, namespaces or any cluster wide and admin controlled objects. The challenge of providing this pattern without tenants information disclosure is shared. Which raise the question of how to provide such GitOps capability to tenants without disclosing information about tenants?

**Reply by stefanprodan** [reply]
> To improve on those restrictions we could imagine decoupling the Sources objects and the reviewing/approving by introducing a SourcePolicy object that defines policies on Sources URLs for tenants (or any other Source attributes). The tk control plane would provide a controller with an admission webhook that is fed by SourcePolicy objects. SourcePolicy objects would be controlled by admin through a central repository open for PR by tenants. Tenants would be responsible to create Source objects in their own namespaces using the tk CLI to sync it with a Git repository. That would improve 1 and 2 limitations but 3 is still present.

My hope is that we could use OPA Gatekeeper for this and not develop our own policy agent. The toolkit could come with a set of `rego` rules packaged as `ConstraintTemplates`. The cluster admins would use the templates to create constraints adapted to their needs (such as allow/deny lists) and commit those constraints to the central repository. This would give admins more flexibility on defining policies not only for Sources but also for enforcing service accounts on Kustomizations, restrict a Kustomization source ref to its namespace, etc.

**Reply by pjbgf** [reply]
@Perdjesk There are a few in-flight RFCs which proposes changes to the current multi-tenancy models and the sharing of resources across tenants. Please take a look and feel free to share your views: https://github.com/fluxcd/flux2/pull/2092 https://github.com/fluxcd/flux2/pull/2093.
